rmw_opendds_cpp/src/rmw_subscription.cpp

```diff
@@ -344,83 +329,19 @@ rmw_subscription_get_actual_qos(
-    return RMW_RET_ERROR;
-  }
-  auto participant = static_cast<DDS::DomainParticipant *>(node_info->participant);
-  if (!participant) {
-    RMW_SET_ERROR_MSG("participant handle is null");
-    return RMW_RET_ERROR;
-  }
-  // TODO(wjwwood): need to figure out when to unregister types with the participant.
-  auto result = RMW_RET_OK;
-  OpenDDSStaticSubscriberInfo * subscriber_info =
-    static_cast<OpenDDSStaticSubscriberInfo *>(subscription->data);
-  if (subscriber_info) {
-    node_info->subscriber_listener->remove_information(
-      subscriber_info->dds_subscriber_->get_instance_handle(), EntityType::Subscriber);
-    node_info->subscriber_listener->trigger_graph_guard_condition();
-    auto dds_subscriber = subscriber_info->dds_subscriber_;
-    if (dds_subscriber) {
-      auto topic_reader = subscriber_info->topic_reader_;
-      if (topic_reader) {
-        auto read_condition = subscriber_info->read_condition_;
-        if (read_condition) {
-          if (topic_reader->delete_readcondition(read_condition) != DDS::RETCODE_OK) {
-            RMW_SET_ERROR_MSG("failed to delete readcondition");
-            result = RMW_RET_ERROR;
-          }
-          subscriber_info->read_condition_ = nullptr;
-        }
-        if (dds_subscriber->delete_datareader(topic_reader) != DDS::RETCODE_OK) {
-          RMW_SET_ERROR_MSG("failed to delete datareader");
-          result = RMW_RET_ERROR;
-        }
-        subscriber_info->topic_reader_ = nullptr;
-      } else if (subscriber_info->read_condition_) {
-        RMW_SET_ERROR_MSG("cannot delete readcondition because the datareader is null");
-        result = RMW_RET_ERROR;
-      }
-      if (participant->delete_subscriber(dds_subscriber) != DDS::RETCODE_OK) {
-        RMW_SET_ERROR_MSG("failed to delete subscriber");
-        result = RMW_RET_ERROR;
-      }
-      subscriber_info->dds_subscriber_ = nullptr;
-    } else if (subscriber_info->topic_reader_) {
-      RMW_SET_ERROR_MSG("cannot delete datareader because the subscriber is null");
-      result = RMW_RET_ERROR;
-    }
-    RMW_TRY_DESTRUCTOR(
-      subscriber_info->~OpenDDSStaticSubscriberInfo(),
-      OpenDDSStaticSubscriberInfo, result = RMW_RET_ERROR)
-    rmw_free(subscriber_info);
-    subscription->data = nullptr;
-  }
-  if (subscription->topic_name) {
-    rmw_free(const_cast<char *>(subscription->topic_name));
-  }
-  rmw_subscription_free(subscription);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info, "node_info is null", return RMW_RET_ERROR);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info->participant.in(), "participant is null", return RMW_RET_ERROR);
 
-  return result;
+  RMW_CHECK_FOR_NULL_WITH_MSG(subscription, "subscription is null", return RMW_RET_ERROR);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(subscription, subscription->implementation_identifier,
+    opendds_identifier, return RMW_RET_ERROR)
+
+  // TODO(wjwwood): need to figure out when to unregister types with the participant.
```

Comment:
I'm not sure the value in retaining these comments. Does it still really make sense?

Reply:
Yes. I had also put a note in the beginning comment of this PR, but forgot to bring it up in the meeting earlier. I'll do it in tomorrow's meeting.

Reply:
@mitza-oci Would you please take a look at this?

Reply:
OpenDDS provides the `unregister_type` method on TypeSupport as an extension to the DDS standard.  You can use it in RMW.